numpy/core/numeric.py

```diff
@@ -2341,6 +2361,11 @@ def within_tol(x, y, atol, rtol):
     x = asanyarray(a)
     y = asanyarray(b)
 
+    if not np.can_cast(x.dtype, y.dtype, casting=casting):
```

Comment:
```suggestion
    if not (
        np.can_cast(x.dtype, y.dtype, casting=casting) or
        np.can_cast(y.dtype, x.dtype, casting=casting)
    ):
```
We'd have to check both combinations, 
otherwise you'd run into the risk of `(int, float)` being allowed while `(float, int)` isn't.
There might be a more elegant way to do this, but I can't think of one right now.

This might get a bit slow though, so I'd say adding a fast-path for `casting == INSERT_DEFAULT_VALUE` would be worthwhile.